File: 강화학습코드예제-gpt.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>

const int NUM_STATES = 5;
const int NUM_ACTIONS = 2;
double Q[NUM_STATES][NUM_ACTIONS] = {0};

double alpha = 0.1;  // 학습률
double gamma = 0.9;  // 할인율
double epsilon = 0.1; // 탐험률

int choose_action(int state) {
    if ((rand() / double(RAND_MAX)) < epsilon)
        return rand() % NUM_ACTIONS;
    return Q[state][0] > Q[state][1] ? 0 : 1;
}

int get_next_state(int state, int action) {
    // 간단한 환경
    return (state + action) % NUM_STATES;
}

double get_reward(int state, int action) {
    return (state == NUM_STATES - 1) ? 1.0 : 0.0;
}

void train(int episodes) {
    for (int ep = 0; ep < episodes; ++ep) {
        int state = rand() % NUM_STATES;
        for (int t = 0; t < 100; ++t) {
            int action = choose_action(state);
            int next_state = get_next_state(state, action);
            double reward = get_reward(state, action);
            Q[state][action] += alpha * (reward + gamma * std::max(Q[next_state][0], Q[next_state][1]) - Q[state][action]);
            state = next_state;
        }
    }
}

File: 강화학습코드예제-gpt.py
import gym
import torch
import torch.nn as nn
import torch.optim as optim
import random
import numpy as np
from collections import deque

# Q-Network 정의
class DQN(nn.Module):
    def __init__(self, state_dim, action_dim):
        super(DQN, self).__init__()
        self.fc = nn.Sequential(
            nn.Linear(state_dim, 128),
            nn.ReLU(),
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Linear(64, action_dim)
        )
    
    def forward(self, x):
        return self.fc(x)

# 환경 설정
env = gym.make('CartPole-v1')
state_dim = env.observation_space.shape[0]
action_dim = env.action_space.n

# 하이퍼파라미터
lr = 0.001
gamma = 0.99
epsilon = 1.0
epsilon_decay = 0.995
epsilon_min = 0.01
batch_size = 64
memory = deque(maxlen=10000)
episodes = 100

# 네트워크와 옵티마이저
q_network = DQN(state_dim, action_dim)
optimizer = optim.Adam(q_network.parameters(), lr=lr)
loss_fn = nn.MSELoss()

# 경험을 이용한 학습 함수
def train():
    if len(memory) < batch_size:
        return
    batch = random.sample(memory, batch_size)
    states, actions, rewards, next_states, dones = zip(*batch)

    states = torch.FloatTensor(states)
    actions = torch.LongTensor(actions).unsqueeze(1)
    rewards = torch.FloatTensor(rewards)
    next_states = torch.FloatTensor(next_states)
    dones = torch.FloatTensor(dones)

    q_values = q_network(states).gather(1, actions).squeeze()
    next_q_values = q_network(next_states).max(1)[0]
    expected_q = rewards + gamma * next_q_values * (1 - dones)

    loss = loss_fn(q_values, expected_q.detach())
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

# 메인 루프
for episode in range(episodes):
    state, _ = env.reset()  # 최신 Gym은 (obs, info) 반환
    total_reward = 0
    done = False
    while not done:
        if random.random() < epsilon:
            action = env.action_space.sample()
        else:
            with torch.no_grad():
                action = q_network(torch.FloatTensor(state)).argmax().item()

        next_state, reward, terminated, truncated, _ = env.step(action)
        done = terminated or truncated

        memory.append((state, action, reward, next_state, float(done)))
        state = next_state
        total_reward += reward
        train()
    
    epsilon = max(epsilon_min, epsilon * epsilon_decay)
    print(f"Episode {episode + 1}, Reward: {total_reward}, Epsilon: {epsilon:.3f}")

env.close()
